Type checking and coercion

tclose opened the pull request:
## Types of changes
<!--- What types of changes does your code introduce? Keep only relevant points: -->

This PR contains a number of breaking changes:
* significantly stricter type-checking
* lists of inputs to be split over, including lazy inputs, now need to set in the `TaskBase.split()` method (i.e. instead of in Task.__init__ or by setting inputs attribute).
* Splits and combines need to be applied to typed tasks/workflows before their `lzout` interface is accessed (so the type of the LazyField can be set properly)
* ShellCommandTask Inputs that use `output_file_template` must also be defined in the output spec with `fileformats.generic.File` (or subclass) type, in addition to a input field of `str`, `Path`, `ty.Union[str, bool]` or  `ty.Union[Path, bool]` type (@ghisvail I forgot about this when you asked whether there were any implications for task interface design).
    * NB: The automatic insertion would work if it was in the other direction than it is currently, i.e. a File-type field defined in the output spec auto inserts a Path-type input field 
* MultiInput(Obj|File) types changed to be proper classes instead of just lists
* MultiOutput(Obj|File) types changed to be a Union beween a single T, list[T], and a dummy MultiOutputType type used to signify that a converter needs to be inserted into to the field
* `set_input_validators()` has been removed
* `File` and `Directory` in `pydra.engine.specs` now refer to the classes in `fileformats.generic`. As such, existence (and format) checking of non-lazy inputs is performed in `FileSet.__init__`, not at runtime as was the case previously
* Default values assigned to fields in input/output specs are now set to `attrs.NOTHING` not None (not sure if this actually changes the overall behaviour though as I believe they are always overwritten)
* Associated files (e.g. JSON side-cars) won't be copied unless an appropriate file-format class is specified (i.e. `File` is not sufficient and you would need to use `fileformats.medimage.NiftiX`). However, on the plus side
    * Non-neuroimaging associated files will be copied if an appropriate format is used (currently only selected neuroimaging formats with associated files are supported, which are hard-coded in helpers_file.py)
    * Associated files will be included in hashes
* No longer compatible with some older micro versions of Python (e.g. 3.8.0) as they don't support typing in quite the same way

## Summary

### Type checking

Adds detailed type checking at both run and construction time (i.e. lazy fields carry the type of the field they point to). This is done by `pydra.utils.TypeParser`, which is works as an `attrs` converter and is set to be the converter into every field in the generated input/output specs by default.

The type parser unwraps arbitrary nested generic types (e.g. List, Tuple and Dict in typing package) and tests that they either match or are a subclass of the declared types.

### Type coercion

In addition to the type-checking, selected types will be automatically coerced to their declared type:

* typing.Sequence -> typing.Sequence (but not str -> typing.Sequence or typing.Sequence -> str)
* typing.Mapping -> typing.Mapping
* Path -> os.PathLike
* str -> os.PathLike
* os.PathLike -> Path
* os.PathLike -> str
* typing.Any -> MultiInputObj
* typing.Union[os.PathLike, str] -> MultiInputFile
* typing.Sequence -> MultiOutputObj
* int -> float
* numpy.array -> typing.Sequence
* numpy.<primitives> -> <python-primitives>

Like the type-checking, the coercion is performed between types within arbitrarily nested containers so can handle coercions, e.g.

`dict[str, list[str]] -> dict[str, tuple[File]]`

### Hashing

Hashing is implemented (by @effigies) in a dispatch method, which handles hashes for non-ordered objects such as sets in a process-independent way.

Hashing of files is now handled outside of Pydra within the `fileformats` package. This enables the hashes to be aware of changes in associated files.

### Copying files

Copying files (e.g. into working directories) is also handled in `fileformats`, and includes graceful fallback between leaving the files in-place, symlinking, hardlinking and copying files (in order of preference), taking into account whether the files need to be in the same folder/have the same file-stem.

## Checklist
<!--- Please, let us know if you need help-->
- [x] I have added tests to cover my changes (if necessary)
- [x] I have updated documentation (if necessary)

satra commented:
@tclose - thanks for all the effort in bringing these different PRs and ideas together. the one thing that struck me was the introduction of `gathered`.  this particular thing seems to be a requirement of type checking rather than readability of code. python isn't really designed to be that kind of language and it seems we are forcing some elements to users here to satisfy a developer constraint. i'm not for or against this at this stage. just wanted to bring it up so that the user perspective is considered. 

a key goal of pydra was to keep it simple for users to not have to do things they don't fully understand. one element of this was the idea that pydra can be used in a jupyter notebook or a script without the need to construct a workflow or define an interface. simply used to parallelize elements of a script.

tclose commented:
> @tclose - thanks for all the effort in bringing these different PRs and ideas together. the one thing that struck me was the introduction of `gathered`. this particular thing seems to be a requirement of type checking rather than readability of code.

Yeah, I tried to avoid the need to add the "gathered" syntax as it seems a bit superfluous, but having to check whether a value is actually a list of that value, or a list of list of that value, etc... just makes the type-checking impractical (unless anyone has any good ideas). I personally think it makes the code a bit easier to understand if values that are to be split are explicitly marked as such, but definitely something it would be good to discuss.

Note that the "type-checking" process is actually as much about coercing strings/paths into File objects so we know how to hash them properly as it is about validation. So I think it will be hard to get away from some form of type-parsing (the existing code this PR replaces already had quite a bit of special-case handling to deal with File objects).

> python isn't really designed to be that kind of language and it seems we are forcing some elements to users here to satisfy a developer constraint. i'm not for or against this at this stage. just wanted to bring it up so that the user perspective is considered.

I would perhaps argue that the direction of travel with Python is to move towards a stricter, type-checked, language. Certainly editors like VSCode are pushing you to write code that way, with type-checked linting switched on by default.

On the topic of linting, I have been thinking about what would it would take to facilitate static checking and hinting of pydra workflows. I will try to put together an issue summarising what would be required so we can also debate its merits. 

> a key goal of pydra was to keep it simple for users to not have to do things they don't fully understand. one element of this was the idea that pydra can be used in a jupyter notebook or a script without the need to construct a workflow or define an interface. simply used to parallelize elements of a script.

Perhaps if it had a better name than `gathered` it would be more intuitive. I played around with a few but couldn't come up with one that I really liked. Personally, I reckon by the time that the users are thinking about using split, they would already be looking up the docs and in that case having to use gathered wouldn't be too much of an extra burden. But I agree it isn't quite as clean.

djarecka commented:
on one hand I don't like introducing like `gathered` (not sure what would be the best name, we can think), on the other hand I'm not sure if it decreases readability, or actually improves the readability... 
It does explicitly points to the argument that should have ben modified when moving the task to the splitting option, so it might help with debugging.

effigies commented on pydra/utils/typing.py:
```diff
@@ -0,0 +1,603 @@
+        self.not_coercible = list(not_coercible) if not_coercible is not None else []
+        self.pattern = expand_pattern(tp)
+
+    def __call__(self, obj: ty.Any) -> T:
```
Going by the docstring, it seems like this is actually:

```suggestion
    def __call__(self, obj: object) -> T | LazyField[T]:
```

1) You don't need `Any`. You're actually saying "all this needs to be is an object, and I'll either convert it or raise an exception". `object` is the correct type for that.
2) I think it's worth explicitly saying this could return a `Future` that emits the correct type (or that we'll convert when appropriate).

(We do still need to special-case `attr.NOTHING` for now, since you would have to use the private `attr._Nothing` type. `Literal[Nothing]` is broken: https://github.com/python-attrs/attrs/pull/983)

This seems like it will probably come up a lot. If so, we would want to represent this concept with a type alias like `Field[T]`:

```Python
Field: TypeAlias = T | LazyField[T]
```

tclose replied:
Ok, will look to tighten that up

effigies commented on pydra/engine/specs.py:
```diff
@@ -738,43 +678,71 @@ class SingularitySpec(ContainerSpec):
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class LazyField:
+    """Lazy fields implement promises."""
+
+    name: str
+    field: str
+    attr_type: str
+    type: ty.Type[ty.Any]
```
Is this possible?

```suggestion
class LazyField(ty.Generic[T]):
    """Lazy fields implement promises."""

    name: str
    field: str
    attr_type: str
    type: ty.Type[T]
```

tclose replied:
I think so yes, just not sure if it adds much.

effigies commented on pydra/engine/specs.py:
```diff
@@ -738,43 +678,71 @@ class SingularitySpec(ContainerSpec):
 
     def __repr__(self):
-        return f"LF('{self.name}', '{self.field}')"
+        return f"LF('{self.name}', '{self.field}', {self.type})"
```
IDK if this would be more readable? Depends on the `repr` of `self.type`. For types

```suggestion
        return f"Lazy[{self.type}]('{self.name}', '{self.field}')"
```

effigies commented:
> With LazyFields, would it look like?
> 
> ```python
> > a = TaskC().split("x", x=wf.b.lzout.out)
> ```

That would make sense to me. I think `TaskC(x=wf.b.lzout.out).split('x')` would be problematic.

For a transition, I think we could easily catch cases in

https://github.com/nipype/pydra/blob/426564eca3efa7d840fdca33bdbef6a1c88e2ffc/pydra/engine/core.py#L169-L172

where the input is a list of the expected type and suggest setting the values in `split()`. This isn't actually a new syntax, but we would need to change the behavior to track these inputs in `.state` (or some `._stateful_inputs`, if we want to keep the `.state` object itself clean).

> > Lazy inputs/outputs are not going to be reasonably type-checkable,
> 
> Do you mean statically by linters?

Yes. And at runtime, by the time we do type checking/coercion it will already have been split along the list, whether that's bare `list` or a subclass/newtype.

tclose commented:
> Thinking a bit about `gathered` and trying to find a way out. I'm worried that a special type will add confusion when we do something like `A -> B -> C` where `A` is mapped over input, `B` accepts a list, and `C` again maps over inputs.

I'm trying to write a test for this case, but I'm not sure I understanding the splitting/combining syntax properly. Is this a valid workflow?

```python
    wf = Workflow(name="test", input_spec=["x", "y"])

    wf.add( # Split over workflow input "x" on "scalar" input
        list_mult_sum(
            in_list=wf.lzin.x,
            scalar=wf.lzin.x,
            name="A",
        ).split("scalar")
    )

    wf.add(  # Workflow is still split over "x"
        list_mult_sum(
            name="B",
            scalar=wf.A.lzout.sum,
            in_list=wf.A.lzout.products,
        )
    )

    wf.add(  # Workflow is combined over "x"
        list_mult_sum(
            name="C",
            scalar=wf.lzin.y,
            in_list=wf.B.lzout.sum,
        ).combine("A.scalar")
    )

    wf.add(  # Workflow is split again, this time over C.products
        list_mult_sum(
            name="D",
            scalar=wf.C.lzout.products,
            in_list=wf.lzin.x,
        ).split("scalar")
    )

    wf.add(  # Workflow is finally combined again into a single node
        list_mult_sum(name="E", scalar=wf.lzin.y, in_list=wf.D.lzout.sum).combine(
            "D.scalar"
        )
    )

    wf.set_output(
        [
            ("alpha", wf.D.lzout.sum),
            ("beta", wf.D.lzout.products)
        ]
    )
```

where 


```python

@mark.task
@mark.annotate({"return": {"sum": int, "products": ty.List[int]}})
def list_mult_sum(scalar: int, in_list: ty.List[int]) -> ty.Tuple[int, ty.List[int]]:
    products = [scalar * x for x in in_list]
    return functools.reduce(operator.add, products, 0), products
```

tclose commented:
> @tclose - i've read the workflow and I think it should work, but it uses something what we called "inner splitter", since it's not really splitting over an input, but an output of a node (that could be a list).
> 
> Does it fail? I could try to run it later

I worked out the error was in the `combine` call, which I had meant to combine the inputs to C, and therefore should have been done on node B

tclose commented:
Note that the coverage actually increases >1% but codecov is confused

tclose commented:
The errors that have just shown up with Ubuntu >= 3.10 appear to be an intermittent failure where the output fields aren't being set properly (i.e. result.output.* of shell tasks sometimes are attr.NOTHING). I believe have run across it before but haven't been able to pin it down due to its intermittent nature.

For example, testing on `Linux ip-10-0-2-164 5.19.0-1027-aws #28~22.04.1-Ubuntu SMP Wed May 31 18:30:36 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux` with Python==3.10.6, I don't get any errors.

It appears to be an issue with multiprocessing, as switching `test_shell_cmd_inputsspec_11` to use the "serial" plugin instead of "cf" resolves the issue and `test_shell_cmd_outputspec_7` and `test_shell_cmd_outputspec_7a` are only failing on the "cf" test parameterization.

djarecka commented:
thanks so much for doing it! will be reviewing!

djarecka commented on pydra/engine/tests/test_node_task.py:
```diff
@@ -769,12 +781,11 @@ def test_task_state_1(plugin_dask_opt, input_type, tmpdir):
     """task with the simplest splitter (inputs set separately)"""
     nn = fun_addtwo(name="NA")
-    nn.split(splitter="a")
-    nn.inputs.a = [3, 5]
```
I understand that setting input later is a still valid option, right?

djarecka replied:
actually, I guess it is not...

tclose replied:
> actually, I guess it is not...

You can but you need to wrap the input in a `StateArray` object, e.g.

`nn.inputs.a = specs.StateArray([3, 5])`

djarecka replied:
so, perhaps we can use it in the test, otherwise `test_task_state_1a` is the same as `test_task_state_1`

tclose commented on pydra/utils/tests/test_typing.py:
```diff
@@ -0,0 +1,599 @@
+            in_file=wf.specific1.lzout.out,
+            name="generic",
+        )
+    )
+
+    with pytest.raises(TypeError, match="Cannot coerce"):
+        # No cast of generic task output to MyFormatX
+        wf.add(
+            specific_task(
+                in_file=wf.generic.lzout.out,
+                name="specific2",
+            )
+        )
+
+    wf.add(
+        specific_task(
+            in_file=wf.generic.lzout.out.cast(MyFormatX),
+            name="specific2",
+        )
+    )
+
+    wf.set_output(
+        [
+            ("out_file", wf.specific2.lzout.out),
+        ]
+    )
+
+    my_fspath = tmp_path / "in_file.my"
+    hdr_fspath = tmp_path / "in_file.hdr"
+    my_fspath.write_text("my-format")
+    hdr_fspath.write_text("my-header")
+    in_file = MyFormatX([my_fspath, hdr_fspath])
+
+    result = wf(in_file=in_file, plugin="serial")
+
+    out_file: MyFormatX = result.output.out_file
+    assert type(out_file) is MyFormatX
+    assert out_file.parent != in_file.parent
+    assert type(out_file.header) is MyHeader
+    assert out_file.header.parent != in_file.header.parent
```
Following on from the discussion last night, I realised that I could get the `LazyField.cast()` method to work for the case @effigies raised with only a minor adjustment. So now the following test passes for all combinations of shell and function tasks

djarecka commented on pydra/engine/tests/test_helpers.py:
```diff
@@ -45,14 +45,14 @@ def test_save(tmpdir):
-        helpers_file.hash_file(outdir / "test.file")
-        == "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"
-    )
+# def test_hash_file(tmpdir):
+#     outdir = Path(tmpdir)
+#     with open(outdir / "test.file", "w") as fp:
+#         fp.write("test")
+#     assert (
+#         helpers_file.hash_file(outdir / "test.file")
+#         == "37fcc546dce7e59585f3217bb4c30299"
+#     )
```
could we keep a similar test?

tclose replied:
👍

djarecka commented on pydra/engine/helpers_file.py:
```diff
@@ -544,31 +51,40 @@ def ensure_list(filename):
-                copyfile(file, newfile, copy=copy)
-                map_copyfiles[fld.name] = str(newfile)
-    return map_copyfiles or None
+def copy_nested_files(
```
perhaps we could have tests for these new functions in test_helpers_file?

tclose replied:
Good idea. I have added them now (and found/fixed a bug in the process)

djarecka commented on pydra/engine/tests/test_nipype1_convert.py:
```diff
@@ -119,4 +119,4 @@ def test_interface_run_1():
     assert task.cmdline == "touch hello.txt"
     res = task()
-    assert res.output.new_file.exists()
+    assert res.output.new_file.fspath.exists()
```
I wonder if `File` could also have `exists` methods?

tclose replied:
fspaths in File are actually already checked for existence, so this check is redundant. I just kept it there so that new_file is checked to be a File essentially. Probably a more elegant way to do this but I was trying to keep the tests as similar as possible

tclose commented on pydra/engine/core.py:
```diff
@@ -597,17 +708,26 @@ def combine(self, combiner, overwrite=False):
                 "combiner has been already set, "
                 "if you want to overwrite it - use overwrite=True"
             )
+        # if kwargs:
+        #     new_inputs = {}
+        #     for inpt_name, inpt_val in kwargs.items():
+        #         if not isinstance(inpt_val, LazyField):
+        #             raise TypeError(
+        #                 "Only lazy-fields can be set as inputs in the combine method "
+        #                 f"not {inpt_name}:{inpt_val}"
+        #             )
+        #         new_inputs[inpt_name] = inpt_val.combine()
+        #     self.inputs = attr.evolve(self.inputs, **new_inputs)
```
👍

tclose commented on pydra/engine/helpers.py:
```diff
@@ -1,37 +1,31 @@
-
-
+import attr
+import attrs  # New defaults
```
Might be a bit cleaner to use `attrs` throughout, but there are no plans to deprecate `attr` so it won't matter

djarecka commented on pydra/engine/core.py:
```diff
@@ -566,24 +617,84 @@ def split(self, splitter, overwrite=False, cont_dim=None, **kwargs):
 
-    def combine(self, combiner, overwrite=False):
+    @classmethod
+    def _unwrap_splitter(
```
oh, ok, this is just to get the list of fields

djarecka commented:
Thank you again for the PR! I think we should  merge it if you don't want to add anything more here! During the OHBM hackathon I will try to work on tutorial updates, so if I have any questions or other suggestions we can do in a separate pr.

tclose commented:
> Thank you again for the PR! I think we should merge it if you don't want to add anything more here! During the OHBM hackathon I will try to work on tutorial updates, so if I have any questions or other suggestions we can do in a separate pr.

Great, thanks!

effigies commented on pydra/engine/tests/test_shelltask.py:
```diff
@@ -2758,30 +2760,38 @@ def test_shell_cmd_outputspec_2a_exception(plugin, tmpdir):
 
+@pytest.mark.xfail(
+    reason=(
+        "This test doesn't look like it ever worked properly. The command isn't being "
+        "split on ';' and instead the arguments are just treated as a list of dirs to create. "
+        "This includes 'tmp/newfile.txt', which fileformats now correctly detects as being "
+        "a directory instead of a file"
```
If this had happened to work for a time, I suppose that's fine, but it doesn't seem worth testing and supporting. I vote for just removing these tests.

effigies reviewed (approved):
Batch of comments. I started out going through core stuff, but then switched over to tests, as I think verifying that we only see expected syntax changes (and bug fixes) is most important. I got through `test_shelltask`, and all my suggestions are minor.

I approve this PR in principle, and if I fail to completely finish my review, please do not block on me.

This is amazing work, Tom! I hope to get through the rest before we talk, but in general I'm really impressed how this has come together in a way that keeps the basic type-free functionality intact.

effigies commented on pydra/engine/tests/utils.py:
```diff
@@ -165,41 +205,43 @@ def identity_2flds(
 
 @mark.task
-def list_sum(x):
+def list_sum(x: ty.Sequence[ty.Union[int, float]]) -> ty.Union[int, float]:
     return sum(x)
```
Have we tested on any generics, like:

```suggestion
NumT = ty.TypeVar('NumT', bound=float)

def list_sum(x: ty.Sequence[NumT]) -> NumT:
    return sum(x)
```

These sorts of functions definitely come up, so it might be nice to make sure we can handle them, though that can be a future PR.

tclose replied:
I have created this issue detailing what would be required to support this, https://github.com/nipype/pydra/issues/672. Shouldn't be too much work, maybe a day or two all going well

effigies commented on pydra/utils/hash.py:
```diff
@@ -0,0 +1,330 @@
+    try:
+        return hash_single(obj, Cache({}))
+    except Exception as e:
+        hash_single(obj, Cache({}))  # for debugging
```
```suggestion
```

effigies commented on pydra/utils/tests/test_hash.py:
```diff
@@ -0,0 +1,215 @@
+
+
+def test_registration_conflict():
+    # Verify the order of
```
```suggestion
    # Verify the order of precedence: class/superclass registration, __bytes_repr__, protocols
```

effigies commented on pydra/utils/hash.py:
```diff
@@ -0,0 +1,330 @@
+        if obj.dtype == "object":
+            yield from bytes_repr_sequence_contents(iter(obj.ravel()), cache)
+        else:
+            bytes_it = iter(obj.tobytes(order="C"))
+            for chunk in iter(
+                lambda: bytes(itertools.islice(bytes_it, NUMPY_CHUNK_LEN)), b""
+            ):
+                yield chunk
```
We're already allocating the RAM, unless there's a speedup I don't understand, I think we should keep this simple.

```suggestion
            yield obj.tobytes(order="C")
```

tclose replied:
Oh, I didn't know that RAM allocation was the rate-limiting factor when trying to hash large objects. I assumed it was algorithmic but it is a bit of a black box to me

effigies reviewed (commented):
Another batch of comments. Through all the tests except test_typing.

tclose commented on pydra/engine/tests/test_helpers.py:
```diff
@@ -210,7 +178,7 @@ def test_load_and_run(tmpdir):
     task_pkl = Path(tmpdir.join("task_main.pkl"))
 
-    task = multiply(name="mult", x=[1, 2], y=10).split("x")
+    task = multiply(name="mult").split("x", x=[1, 2], y=10)
```
yep, that is possible. Have updated it

tclose commented on pydra/engine/tests/test_node_task.py:
```diff
@@ -176,9 +177,9 @@ def test_task_init_4b():
 
 def test_task_init_4c():
     """trying to set splitter twice without using overwrite"""
-    nn = fun_addtwo(name="NA").split(splitter="b", a=[3, 5])
+    nn = fun_addvar(name="NA").split(splitter="b", b=[1, 2])
     with pytest.raises(Exception) as excinfo:
-        nn.split(splitter="a")
+        nn.split(splitter="a", a=[3, 5])
     assert "splitter has been already set" in str(excinfo.value)
```
Yes, that would be good and not difficult to implement I reckon

tclose replied:
I have created a feature request for this, https://github.com/nipype/pydra/issues/680

tclose commented on pydra/engine/tests/test_submitter.py:
```diff
@@ -575,6 +573,7 @@ def test_sge_no_limit_maxthreads(tmpdir):
     assert job_1_endtime > job_2_starttime
 
 
+@pytest.mark.xfail(reason="Not sure")
```
Ah yes, this is that test I wrote to hit the exception handling code (which attempts to help find where the problem is). Not sure why it would no longer work, maybe we are copying the inputs somewhere

tclose replied:
Got it, just needed to implement the __bytes_repr__ to A. Not sure if it points to a case that we need to handle better with the hashing functions though.

tclose commented on pydra/engine/tests/test_tasks_files.py:
```diff
@@ -94,7 +94,7 @@ def test_wf_1(tmpdir):
     wf.inputs.file_orig = file_orig
 
-    with Submitter(plugin="cf") as sub:
+    with Submitter(plugin="serial") as sub:
```
It isn't actually failing. I just find serial is easier to use when debugging. I think I just left this here for a bit of variety but have reverted it back

tclose commented on pydra/engine/core.py:
```diff
@@ -223,10 +229,12 @@ def __setstate__(self, state):
-        if name == "lzout":  # lazy output
-            return LazyField(self, "output")
-        return self.__getattribute__(name)
+    @property
+    def lzout(self):
+        if self._lzout:
+            return self._lzout
+        self._lzout = LazyOut(self)
+        return self._lzout
```
Nice, wasn't aware of that one!

tclose commented on pydra/engine/core.py:
```diff
@@ -356,7 +358,9 @@ def generated_output_names(self):
         output_klass = make_klass(self.output_spec)
         if hasattr(output_klass, "generated_output_names"):
-            output = output_klass(**{f.name: None for f in attr.fields(output_klass)})
+            output = output_klass(
+                **{f.name: attr.NOTHING for f in attr.fields(output_klass)}
+            )
```
This works around that some fields might be mandatory (i.e. no default value). Whether any field should be marked as mandatory is another matter I suppose

tclose commented on pydra/engine/core.py:
```diff
@@ -534,28 +555,58 @@ def _run(self, rerun=False, **kwargs):
         run_output = self.output_
         output_klass = make_klass(self.output_spec)
-        output = output_klass(**{f.name: None for f in attr.fields(output_klass)})
+        output = output_klass(
+            **{f.name: attr.NOTHING for f in attr.fields(output_klass)}
+        )
```
see above

tclose commented on pydra/engine/core.py:
```diff
@@ -534,28 +555,58 @@ def _run(self, rerun=False, **kwargs):
+            a reference to the task
         """
+        if self._lzout:
+            raise Exception(
```
I don't think Pydra defines its own errors. I've heard this is the current trend to just use standard errors, which I kind of get but feel like you are throwing away some useful information at times. Will change to RuntimeError

tclose commented on pydra/engine/core.py:
```diff
@@ -534,28 +555,58 @@ def _run(self, rerun=False, **kwargs):
+            splitter = list(inputs)
+        elif splitter:
+            missing = set(self._unwrap_splitter(splitter)) - set(inputs)
+            missing = [m for m in missing if not m.startswith("_")]
```
`task.split(splitter="[(a, b), c]", a=[1, 2, 3], b=[4, 5, 6])` (c is missing)

tclose replied:
ah sorry, misread your comment. The "_" prefix is used to refer to upstream split tasks I believe (the splitting documentation needs a fair bit more work as I discovered a fair bit of undocumented features doing this PR)

tclose commented on pydra/engine/tests/test_shelltask.py:
```diff
@@ -1589,31 +1592,34 @@ def test_shell_cmd_inputsspec_11():
     wf.set_output([("out", wf.echoMultiple.lzout.outputFiles)])
 
-    with Submitter(plugin="cf") as sub:
+    with Submitter(plugin="serial") as sub:
```
Again, no failure, just left as serial for diversity. Reverted

tclose commented on pydra/engine/tests/test_shelltask.py:
```diff
@@ -3163,13 +3186,13 @@ def test_shell_cmd_outputspec_7a(tmpdir, plugin, results_function):
     )
 
-    res = results_function(shelly, plugin)
+    res = results_function(shelly, "serial")
```
cf works, sorry this was just for debugging

tclose replied:
Actually, I had forgotten about this. In other tests I had just left the plugin as serial but in this case "cf" is causing this. The issue is described [here](https://github.com/nipype/pydra/issues/671)

tclose commented on pydra/engine/tests/test_task.py:
```diff
@@ -378,7 +380,7 @@ def testfunc(a: MultiInputObj):
 
     funky = testfunc(a=3.5)
-    assert getattr(funky.inputs, "a") == [3.5]
+    assert getattr(funky.inputs, "a") == MultiInputObj([3.5])
```
Yes, I think I changed this for an earlier iteration, when MultiInputObj was a separate class

tclose commented on pydra/engine/tests/test_workflow.py:
```diff
@@ -2217,13 +2251,11 @@ def test_wfasnd_ndst_updatespl_1(plugin, tmpdir):
+    wfnd.add2.split("x", x=[2, 4])
     wfnd.set_output([("out", wfnd.add2.lzout.out)])
-    # TODO: without this the test is failing
-    wfnd.inputs.x = [2, 4]
 
     wf = Workflow(name="wf", input_spec=["x"])
     wf.add(wfnd)
-    wfnd.add2.split("x")
     wf.set_output([("out", wf.wfnd.lzout.out)])
```
This won't work because, you need to provide a value for x in the line

```python
existing_wf.add2.split('x')
```

I think it might also fail if any of the fields were typed and connected to downstream nodes in the existing_wf as they would be expecting a single value not a state-array.

However, I'm not sure of the use case for this test. When would you want to go into the inner workflow and split only one node and not an input?

If your inner workflow has A->B->C, then you wouldn't be able to split either B or C as the inputs coming from A and B won't be lists. Even if it is on an input which is connected to both A & B, you wouldn't be able to provide a list to one and a single value to the other. So I can't think of a case where would work unless your tasks have inputs of type `Union[list[T], T]`.

djarecka replied:
I'm not sure if we ever had very specific example for it... Just wanted to be possible, but probably you're right that there are not many use cases...

tclose commented:
> This is amazing work, Tom! I hope to get through the rest before we talk, but in general I'm really impressed how this has come together in a way that keeps the basic type-free functionality intact.

Thanks 🙏 I'm really chuffed you think so 😊

On the subject of keeping the basic type-free functionality intact, one alteration I have been considering that would be good to discuss in the meeting, is that of relaxing the type-checking at construction time (i.e. lazy fields) so that base classes can be passed to sub-classed type fields to avoid need to add cast statements between loosely-typed tasks and tightly-typed tasks. Stricter type-checking/coercing can still be performed at runtime.

For example, if B is a subclass of A, then currently you can connect a B-typed output into a A-typed input but if you wanted to connect A-typed to B-typed you would need to cast it to B first. While this aligns with how traditional type-checking is done, we might want to be a bit more flexible so you could connect generic `File` type fields to `Nifti` fields for example. If both upstream and downstream nodes use to specific formats that don't match, e.g. MrtrixImage -> Nifti then we would still raise an error.

It would somewhat reduce the effectiveness of the type-checking, but would probably avoid any (false-positive) cases where the type-checking could be annoying

ghisvail commented on pydra/engine/core.py:
```diff
@@ -566,24 +615,84 @@ def split(self, splitter, overwrite=False, cont_dim=None, **kwargs):
+                for n in self._unwrap_splitter(splitter)
+                if not n.startswith("_")
+            )
+            for inpt_name, inpt_val in inputs.items():
```
Just a comment, but I really feel uncomfortable with weird variable name shortcuts such as `fld`, `mdata` or `inpt`. IMO, it adds unnecessary cognitive load to save a limited amount of keystrokes.

tclose replied:
Yes, it isn't my normal style, I just tried to keep it as close to the original as possible

ghisvail commented on pydra/engine/core.py:
```diff
@@ -629,26 +737,22 @@ def _extract_input_el(self, inputs, inp_nm, ind):
-            inputs_dict = {inp: getattr(self.inputs, inp) for inp in self.input_names}
-            return None, inputs_dict
+        # TODO: doesn't work properly for more cmplicated wf (check if still an issue)
+        input_ind = self.state.inputs_ind[ind]
+        inputs_dict = {}
+        for inp in set(self.input_names):
+            if f"{self.name}.{inp}" in input_ind:
+                inputs_dict[inp] = self._extract_input_el(
```
Follow-up to my comment above, now we've got `inp`, `inpt`, `input` and `inputs`. Then `inp_nm`, `input_ind`, ...

It's getting very difficult for me to keep track of variable meaning in my limited short-term memory.

tclose replied:
I typically use inpt to differentiate from the inbuilt function

ghisvail commented on pyproject.toml:
```diff
@@ -12,7 +12,10 @@ dependencies = [
     "cloudpickle >=2.0.0",
     "etelemetry >=0.2.2",
     "filelock >=3.0.0",
+    "fileformats >=0.6",
```
Do we need an upper bound in case of future API breakage (`<0.7` or `<1`)?

tclose replied:
Not sure, could be useful as fileformats is still evolving but then I don't have much experience with upper bounds and how they play out in practice.

ghisvail replied:
You can use `fileformats >=0.6,<0.7` if you don't guarantee API breakages between `0.x`versions. `fileformats >=0.6,<1` may be useful if you allow yourself API breakages right before committing to version 1.

tclose replied:
Ok, I have done >=0.8,<0.9 just in case for now.

ghisvail reviewed (approved):
I went throught this review "diagonally" and provided minor comments along the way.

tclose commented on pydra/engine/core.py:
```diff
@@ -566,24 +617,84 @@ def split(self, splitter, overwrite=False, cont_dim=None, **kwargs):
-    def combine(self, combiner, overwrite=False):
+    @classmethod
+    def _unwrap_splitter(
+        cls, splitter: ty.Union[str, ty.List[str], ty.Tuple[str, ...]]
```
I think the idea was that with the current syntax you can have composite splits such as [('a', 'b'), 'c'] that are both inner and outer. However, you could make this more explicit by using specialised types, e.g. split.outer(split.inner('a', 'b'), 'c') at the expense of more verbosity. This is something for a different PR though I imagine

tclose commented on pydra/engine/core.py:
```diff
@@ -566,24 +617,84 @@ def split(self, splitter, overwrite=False, cont_dim=None, **kwargs):
+        if isinstance(splitter, str):
+            return [splitter]
+        else:
+            return itertools.chain(*(cls._unwrap_splitter(s) for s in splitter))
```
Ok, I have moved it into helpers_state

tclose commented on pydra/engine/core.py:
```diff
@@ -808,51 +932,66 @@ def _sanitize_input_spec(
+                )
             return SpecInfo(
-                name="Inputs",
+                name=spec_name,
```
Shouldn't spec_name always be a string?

ghisvail replied:
It's Python at the end of the day, so `spec_name` can be anything including `None`.

tclose replied:
I think we can probably just raise the error in this case

tclose commented on pydra/engine/core.py:
```diff
@@ -924,17 +1070,24 @@ def __init__(
         # propagating rerun if task_rerun=True
         self.propagate_rerun = propagate_rerun
 
+    @property
```
I went to do this and realised that despite being stored the property isn't actually being cached in the current code. Not sure whether it should be or not (I'm thinking not). Checking that the checksum of a node doesn't change during execution is something I think we catch explicitly otherwise Pydra can't find the results

ghisvail replied:
I am confused whether your comment means `lzin` could become a `cached_property` or not.

tclose replied:
ah yes, in my VS code plugin your comment was showing up on a different property. I have changd it now in my branch

tclose commented on pydra/engine/tests/test_helpers.py:
```diff
@@ -270,9 +238,7 @@ def test_load_and_run_wf(tmpdir):
-    wf.split("x")
-    wf.inputs.x = [1, 2]
-    wf.inputs.y = 10
+    wf.split("x", x=[1, 2], y=10)
```
done

tclose commented on pydra/engine/tests/test_helpers.py:
```diff
@@ -270,9 +239,7 @@ def test_load_and_run_wf(tmpdir):
-    wf.split("x")
-    wf.inputs.x = [1, 2]
-    wf.inputs.y = 10
+    wf.split("x", x=[1, 2], y=10)
```
does it look ok now?

tclose replied:
I have moved y=10 into the wf init now, is that ok?

tclose commented on pydra/engine/core.py:
```diff
@@ -534,28 +555,58 @@ def _run(self, rerun=False, **kwargs):
+        if self._lzout:
+            raise Exception(
+                f"Cannot split {self} as its output interface has already been accessed"
+            )
```
If you now try to split the node after you have accessed the lzout of the node, it will raise an error, e.g.

```python
wf.add(A(x=1, name="a"))
wf.add(B(w=wf.a.lzout.z))
wf.a.split(y=[1, 2, 3])  # <--- Will raise an error, because it will change z from a single value to a list/state-array
```

tclose commented on pydra/engine/specs.py:
```diff
@@ -738,75 +678,348 @@ class SingularitySpec(ContainerSpec):
+        value : Any
+            the resolved value of the lazy-field
+        """
+        from ..utils.typing import TypeParser  # pylint: disable=import-outside-toplevel
```
It creates a circular import unfortunately
